### PR TITLE
Add LifetimeParam resolving to Semantics

### PR DESCRIPTION
**Pull request by Veykril**
This is stuff required for the lifetime references/definitions PR. I pulled this out to make it easier to review as well as because there is one thing that still has to be addressed which can be found in the review comments.

### CODE REVIEW

File: crates/hir_def/src/generics.rs [hunk 1]
```diff
@@ -340,27 +346,29 @@ impl GenericParams {
...
-impl HasChildSource for GenericDefId {
-    type ChildId = LocalTypeParamId;
-    type Value = Either<ast::Trait, ast::TypeParam>;
-    fn child_source(&self, db: &dyn DefDatabase) -> InFile<SourceMap> {
-        let (_, sm) = GenericParams::new(db, *self);
-        sm
+impl GenericDefId {
+    // FIXME: this shouldn't exist?
+    pub fn child_source(&self, db: &dyn DefDatabase) -> InFile<SourceMaps> {
+        GenericParams::new(db, *self).1
```
Comment (Veykril) [inline crates/hir_def/src/generics.rs:352]: Unfortunately `HasChildSource` is actually being used by `TypeParam`(and for now also `LifetimeParam`) in their `HasSource` impls(can be seen in the diffs here), but as explained in this [zulip channel](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/.236787.20lifetime.20reference.20search.20support) `HasChildSource` can't be used anymore here. Are there better approaches to solve this problem here?

File: crates/hir_def/src/generics.rs [hunk 2]
```diff
@@ -72,7 +72,12 @@ pub enum WherePredicateTypeTarget {
     // FIXME: ForLifetime(Vec<LifetimeParamId>, TypeRef)
 }
 
-type SourceMap = ArenaMap<LocalTypeParamId, Either<ast::Trait, ast::TypeParam>>;
+#[derive(Default)]
+pub struct SourceMaps {
+    // FIXME this shouldn't be public
```
Comment (Veykril) [inline crates/hir_def/src/generics.rs:77]: See Fixme below

File: crates/hir_def/src/generics.rs (hunk 1 above)
Comment (Veykril) [inline crates/hir_def/src/generics.rs:352]: Unfortunately `HasChildSource` is actually being used by `TypeParam`(and for now also `LifetimeParam`) in their `HasSource` impls(can be seen in the diffs here), but as explained in this [zulip channel](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/.236787.20lifetime.20reference.20search.20support) `HasChildSource` can't be used anymore here. Are there better approaches to solve this problem here?

**Comment by kjeremy** [comment]
Can we add some tests?

**Comment by Veykril** [comment]
I don't think this PR has much extra to test aside from maybe the new `Semantics` method, but I don't think `Semantics` has any tests in this regard at all? The code that will test this will follow up on this PR.

File: crates/hir/src/semantics.rs [hunk 3]
```diff
@@ -173,6 +178,10 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         self.imp.descend_node_at_offset(node, offset).find_map(N::cast)
     }
 
+    pub fn resolve_lifetime_param(&self, lifetime_token: &SyntaxToken) -> Option<LifetimeParam> {
```
Comment (matklad) [inline crates/hir/src/semantics.rs:181]: Hm, this definitely needs a `FIXME`, as this really should take an `ast::Something`. We can use a token or a node here. Tokens are specified in https://github.com/rust-analyzer/rust-analyzer/blob/8cba4231092b0d9ca44ea1f545856239e2fb6eed/xtask/src/codegen/gen_syntax.rs#L508-L511, nodes in the ungrammar repo. 

I suggest maybe starting with this untyped structure, and then refactor to `ast::` later, as this ordering should give us more context.

File: crates/hir/src/semantics.rs [hunk 4]
```diff
@@ -392,16 +401,43 @@ impl<'db> SemanticsImpl<'db> {
             .kmerge_by(|node1, node2| node1.text_range().len() < node2.text_range().len())
     }
 
+    fn resolve_lifetime_param(&self, lifetime_token: &SyntaxToken) -> Option<LifetimeParam> {
```
Comment (matklad) [inline crates/hir/src/semantics.rs:404]: ditto

File: crates/hir_def/src/generics.rs (hunk 2 above)
Comment (matklad) [reply crates/hir_def/src/generics.rs:77]: keeping stuff below the facede `hir` crate public is generally OK.

**Comment by Veykril** [comment]
bors r=matklad

**Comment by lnicola** [comment]
Nb. this regressed the memory usage by 14/970 MB, but that's probably unavoidable?

image

**Comment by matklad** [comment]
To clarify, this needs extra 14 MB for the new total of 970?

Yeah, I think we need to pay something for this, but I also wonder if we can add some smart smallvec-like optimizations here...

**Comment by lnicola** [comment]
No, it went from 970 to 984 MB.

**Pull request merged**